### PR TITLE
JENKINS-47225: add proc starter cmdenvs and listener to executing code

### PR DESCRIPTION
**Pull request by MattLud**
This sort of fixes another issue, JENKINS-42582, that existed when using SSHAgent and should fix any other plugins that use procstarter to carry out commands on containers. The plugins expect to see output from the exec of their provided commands, but the k8s plugin doesn't provide a response. they end up not seeing results from their command and misbehaving, even though they get a return code stating otherwise. 

Another issue is that the cmd environment does not reinject the provided variables. Plugins should address this by injecting their environment variables properly back into Jenkins, but, for now, we should still provide support via the provided cmd envs. 

This PR addresses these by adding tee support for the outputstream and exporting the environment variables similar to other environment ones.  The environment variables all use different export approaches though - for instance, CmdEnvs are not keypairs but rather qouted strings. 

One issue that cropped up was that when executing `sh` commands, we would get the jnlp agent's injected environment variables as well, causing obvious problems such as JAVA_HOME being overwritten. The unsatisfying answer was to check for the presence of JENKINS_HOME in the cmdenvs and skip if present.

### CODE REVIEW

**Comment by carlossg** [comment]
Thanks, how could a test check that this is correct and does not regress?

**Comment by MattLud** [comment]
Not quite sure - it was something I was hoping to get some suggestions on.  You could use the ssh-agent for testing (doesn't actually need to ssh into anything) or use a dummy plugin that'd need to manually be loaded.

**Comment by carlossg** [comment]
I get this without and with the patch. Is the `Identity added:` a good enough check that it does work ?

```
Executing shell script inside container [ssh-client] of pod [jenkins-slave-1jtn3-qwfrw]
Executing command: "ssh-agent" 
printf "EXITCODE %3d" $?; exit
SSH_AUTH_SOCK=/tmp/ssh-qsdXYRNYr24A/agent.16; export SSH_AUTH_SOCK;
SSH_AGENT_PID=17; export SSH_AGENT_PID;
echo Agent pid 17;
EXITCODE   0[ssh-agent]   Java/JNR ssh-agent
[ssh-agent] Registered BouncyCastle on the remote agent
[ssh-agent] Started.
[Pipeline] {
[Pipeline] sh
[sshagent] Running shell script
+ ssh -vT -o StrictHostKeyChecking=no git@github.com
```

```
Executing shell script inside container [ssh-client] of pod [jenkins-slave-jrpc2-m0m6g]
Executing command: "ssh-agent" 
printf "EXITCODE %3d" $?; exit
SSH_AUTH_SOCK=/tmp/ssh-Yu4HmQwvcuUn/agent.15; export SSH_AUTH_SOCK;
SSH_AGENT_PID=16; export SSH_AGENT_PID;
echo Agent pid 16;
EXITCODE   0SSH_AUTH_SOCK=/tmp/ssh-Yu4HmQwvcuUn/agent.15
SSH_AGENT_PID=16
Executing shell script inside container [ssh-client] of pod [jenkins-slave-jrpc2-m0m6g]
Executing command: "ssh-add" "/home/jenkins/workspace/sshagent@tmp/private_key_8127245054527851129.key" 
printf "EXITCODE %3d" $?; exit
Identity added: /home/jenkins/workspace/sshagent@tmp/private_key_8127245054527851129.key (/home/jenkins/workspace/sshagent@tmp/private_key_8127245054527851129.key)
Identity added: /home/jenkins/workspace/sshagent@tmp/private_key_8127245054527851129.key (/home/jenkins/workspace/sshagent@tmp/private_key_8127245054527851129.key)
EXITCODE   0EXITCODE   0[ssh-agent] Started.
[Pipeline] {
[Pipeline] sh
[sshagent] Running shell script
+ ssh -vT -o StrictHostKeyChecking=no git@github.com
```

**Comment by carlossg** [comment]
I have added a test that fails on master and works on the branch based on my previous comment, let me know if it looks correct

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -116,18 +117,29 @@ public Launcher decorate(final Launcher launcher, final Node node) {
             public Proc launch(ProcStarter starter) throws IOException {
                 boolean quiet = starter.quiet();
                 FilePath pwd = starter.pwd();
-                String[] commands = getCommands(starter);
 
-                return doLaunch(quiet, pwd, commands);
+                String [] cmdEnvs = starter.envs();
+
+                //check if the cmd is sourced from Jenkins, rather than another plugin; if so, skip cmdEnvs as we are getting other environment variables
+                for(String cmd : cmdEnvs){
```
[the remark this answers is not in the document]
Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Amended and pushed.

**Comment by carlossg** [comment]
Your push removed my test, pushed it again and merged master

**Comment by MattLud** [comment]
Sorry about that! Forgot Github allows maintainers to push into PR branches. 

Only feedback on the test is to use an encrypted key to test  that the ssh passphrase(or any sensitive info passed to the executor) isn't logged by the executor inadvertently.  

I've modified the test to use that if you want me to go ahead and push it.

**Comment by carlossg** [comment]
whatever improves the test!

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 2]
```diff
@@ -205,14 +222,23 @@ public void onClose(int i, String s) {
...
+
                     }
 
                     if (environmentExpander != null) {
                         EnvVars envVars = new EnvVars();
                         environmentExpander.expand(envVars);
                         for (Map.Entry<String, String> entry : envVars.entrySet()) {
                             watch.getInput().write(
-                                    String.format("export %s=\"%s\"%s", entry.getKey(), entry.getValue(), NEWLINE).getBytes(StandardCharsets.UTF_8));
+                                    String.format("export %s=\"%s\"%s", entry.getKey(), StringEscapeUtils.escapeJavaScript(entry.getValue()), NEWLINE).getBytes(StandardCharsets.UTF_8));
```
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: This is from #218.

Either we should solve both problems in one PR, or keep them strictly separated..

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 3]
```diff
@@ -205,14 +222,23 @@ public void onClose(int i, String s) {
...
+                        }
+                    }
+
+                    //setup specific command envs passed into cmd
+                    if (cmdEnvs != null) {
+                        for (String cmdEnv : cmdEnvs) {
+                            watch.getInput().write(
+                                    String.format("export \"%s\"%s", StringEscapeUtils.escapeJava(cmdEnv), NEWLINE).getBytes(StandardCharsets.UTF_8));
```
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: This has the same problems as PR #218.

I suggest to solve both problems there, and close this one.

Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Yeah - I had intended to drop that change from this PR or merge this one in after that, repulling to use the latest env setup.

Comment (marvinthepa) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: I merged this branch into your #218 branch to test locally.

Should I push that?

Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: One thing to note - the cmdenv variables come to us as a string, not a set of envVars. We might need to test it with breaking the value out and escaping it individually.

Comment (marvinthepa) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Thanks, but there is an easy solution for that using the Jenkins API:

```
                    EnvVars envVars = new EnvVars();

                    if (cmdEnvs != null) {
                        for (String cmdEnv : cmdEnvs) {
                            envVars.addLine(cmdEnv);
                        }
                    }
```

Just checking how I can test that right now, as `cmdEnvs` has been empty for all my previous manual tests..

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 4]
```diff
@@ -116,18 +118,29 @@ public Launcher decorate(final Launcher launcher, final Node node) {
...
                 boolean quiet = starter.quiet();
                 FilePath pwd = starter.pwd();
-                String[] commands = getCommands(starter);
 
-                return doLaunch(quiet, pwd, commands);
+                String [] cmdEnvs = starter.envs();
+
+                //check if the cmd is sourced from Jenkins, rather than another plugin; if so, skip cmdEnvs as we are getting other environment variables
+                for (String cmd : cmdEnvs) {
+                    if (cmd.contains(JENKINS_HOME)) {
```
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Mmh, can we be more specific here?

Another environment variable containing the string "JENKINS_HOME" might exist..

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 3 above)
Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Rerunning the existing test @carlossg provided should confirm it works just as well.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 4 above)
Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: The root of the issue is related not being able to determine the class that requested the proct  - when running a simple sh, you get the masters variables injected with no real way to tell you to ignore. If it's a plugin like the ssh-agent, you only see what cmdenvs that were passed along by the that class, which wouldn't include JENKINS_HOME.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 3 above)
Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: I'll also open a new PR once the environment variables change set is merged and close this one.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 4 above)
Comment (marvinthepa) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Mmh, can we be more specific here?

Another environment variable containing the string "JENKINS_HOME" might exist..

I guess it should at least be `cmd.startsWith(JENKINS_HOME+"=")`.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 3 above)
Comment (marvinthepa) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: > I'll also open a new PR once the environment variables change set is merged and close this one.

Sounds like a better idea then merging this PR into #218.

When you do the merge, please note that you can just do

```java
                    EnvVars envVars = new EnvVars();
                    if (environmentExpander != null) {
                        environmentExpander.expand(envVars);
                    }

                    //setup specific command envs passed into cmd
                    if (cmdEnvs != null) {
                        for (String cmdEnv : cmdEnvs) {
                            envVars.addLine(cmdEnv);
                        }
                    }

                    this.setupEnvironmentVariable(envVars, watch);
```

Comment (MattLud) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Will do - I'll close this one out and make those changes.

**Pull request closed without merging**